### ISSUE TITLE
sgesvd_bufferSize int32 overflow with CUDA 10.1

### ISSUE THREAD

**Issue by econtal** [cupy/cupy#2351]
After upgrading from CUDA 9.0 to CUDA 10.1, I noticed I'm not able to compute svd on big matrices because of a int32 overflow in sgesvd_bufferSize here: https://github.com/cupy/cupy/blob/master/cupy/linalg/decomposition.py#L257
When bufferSize is just above `2**31 = 2147483648`, then `sgesvd_bufferSize` fails with `CUSOLVER_STATUS_INVALID_VALUE`, likely because of a wrong cast to negative values. If you continue increasing, you can get positive values again, but wrong ones. See graph below.

This might be an issue with cuSOLVER itself, not cupy, but since I'm not familiar with testing CUDA without cupy I can't tell.
It might be related to #1365 as well, but I haven't tested on CUDA 9.1 nor 10.0.

```python
import numpy
from cupy.cuda import cusolver
from cupy.cuda import device

handle = device.get_cusolver_handle()

def test(m):
  try:
    return cusolver.sgesvd_bufferSize(handle, m, 1)
  except:
    return numpy.nan
values = [(m, test(m)) for m in numpy.linspace(1, 150000, 200).astype('int')]    
```

![sgesvd_buffersize_overflow](https://user-images.githubusercontent.com/3266303/62173044-6e599700-b2e9-11e9-86dc-e177ba6f0225.png)

```
CuPy Version          : 6.2.0
CUDA Build Version    : 10010
```

**Comment by toslunar** [comment]
@kmaehashi and I confirmed that `cupy.linalg.svd(cupy.random.randn(50000, 10), full_matrices=False)` runs with CUDA 9.0 but not with CUDA 10.0.

**Comment by econtal** [comment]
In my opinion the first issue is on CUDA's side, because the memory required to perform an SVD on tall matrices should be linear with respect to the input size, and here it is clearly quadratic. I posted about this in nvidia developer channel (https://devtalk.nvidia.com/default/topic/1057891/gpu-accelerated-libraries/cusolver-can-t-compute-svd-on-tall-matrices-with-cuda-10-1-buffersize-grows-quadratically/) but haven't got any answer yet.

The second issue is the integer overflow. If CUDA's `gesvd` actually requires a quadratic amount of memory, then for values above `2**31`, either we will get `CUSOLVER_STATUS_INVALID_VALUE` (overflow to negative), or we won't get the error but likely won't allocate enough memory (overflow to positive).

**Comment by emcastillo** [comment]
cusolverDnSgesvd_bufferSize was linear in cuda 9.0?
I guess that they changed the allocation algorithm and together with the `int` return type it blowed up.

**Comment by econtal** [comment]
Indeed, with previous version of CUDA `cusolverDnSgesvd_bufferSize` was strictly linear (actually `cusolver.sgesvd_bufferSize(handle, m, 1) == m+224`) 

After doing more tests, I confirm the overflow is still here, but you would need tremendous matrices to reach it.

Here is on CUDA9.2 with the same function `test` as before:
```python
>>> test((1<<31)-224-1) 
2147483647
>>> test((1<<31)-224)
-2147483648
>>> test((1<<31)-1)
-2147483553
>>> test(1<<31)
nan
```

This suggests the issue is not cupy at all, but a wrong logic in CUDA10's `cusolverDnSgesvd_bufferSize`; and that the overflow has always been there but never occurred in practical cases before this logic error.


![image](https://user-images.githubusercontent.com/3266303/62595412-172c6700-b893-11e9-9602-accb04afd533.png)

**Comment by anaruse** [comment]
Please wait for a while. This issue is being inquired to library team.

**Comment by anaruse** [comment]
Thanks for your information again.
This issue is recognized by the library team now. This issue (requiring much more work space) seems to be a side effect of performance improvement in CUDA 10.1 from old version.

**Comment by emcastillo** [comment]
@anaruse  thanks a lot!
I guess that the library team will release a hot-fix for this :).

**Comment by jakirkham** [comment]
@anaruse, is this still an issue with the recent updates applied to CUDA 10.1?

cc @pentschev (for awareness)

**Comment by anaruse** [comment]
Yes, this is still an issue even with CUDA 10.1 Update 2.

**Comment by leofang** [comment]
@anaruse Is this fixed in CUDA 11.x?

**Comment by anaruse** [comment]
This issue should have been fixed in CUDA 11.0.
